Seamless usage when running within Jest

tdd opened the pull request:
Hi all!

Thanks a ton for this module, it's super cool.

However, its existing behavior entirely focuses on people using snapshots *outside of Jest*, but a significant share of users *still use Jest*: they just favor Chai's assertion syntax and plugin ecosystem.

I do a lot of JS training, and this is what most of my trainees tell me. That's my situation, too.

This PR makes it so that, when running in Jest, we just delegate to Jest's built-in snapshot capability. No configuration, no setup cost, just registering the Chai plugin, like any other. *As seamless as can be.*

Tests, examples and docs are updated. I strived to follow the coding style I inferred from the source (even if I'm more of StandardJS fan myself 😉 ).

I hope you like it! Feel free to push back if you have any questions or want tweaks.

This addresses the one reason why I don't use this module in my trainings yet.

FWIW, I'm considering making it even nicer by automatically detecting Enzyme wrappers and then using `enzyme-to-json` to streamline the testing experience even more. This could be done for all modes, too, and be a nice one-up over Jest's built-in snapshotting. I'd love feedback on the idea.

Best,

tdd commented:
Hey @suchipi! Would love your feedback on this.

Best,

suchipi commented:
Hey! This looks interesting. My recommendation up until now has been for people to do something like [this](https://medium.com/@RubenOostinga/combining-chai-and-jest-matchers-d12d1ffd0303), but I can see how this would be useful for easy setup. I'll pull this down and give it a try later this week.

I don't want to automatically add `enzyme-to-json` because jest snapshots are used for more than just React and I want the scope of this module to be relatively small. However, snapshot serializers are a good way to automatically add functionality like that- a quick google search finds [jest-serializer-enzyme](https://github.com/rogeliog/jest-serializer-enzyme). I'd have no problem adding a note to the README with instructions on how to add it, assuming it works- I haven't given it a try yet.

tdd commented:
Hey there,

Re: Enzyme serialization: indeed, registering a serializer would make the most sense, that's a good point. Will adjust my own practice on this, and abandon the "too-smart built-in" approach 😉

tdd commented:
Also re: Jest+Chai aliasing Jest's `expect`: yes, this is what I used to do, but having to switch between expectors and assertion syntaxes (`jestExpect(x).toMatchSnapshot()` vs. `expect(x).to.matchSnapshot()`) is just not a great developer experience. I like consistency 😬

suchipi commented on src/buildMatchSnapshot.js:
```diff
@@ -33,4 +42,19 @@ const buildMatchSnapshot = (utils, parseArgs) => {
 
+const safeRequireJestExpect = () => {
+  try {
+    return require("jest-matchers");
```
Jest is [planning to rename `jest-matchers` to `expect`](https://github.com/facebook/jest/issues/1679#issuecomment-282478002). Could we return the global `expect` here instead?
```js
if (typeof expect !== "undefined") {
  return expect;
} else {
  return null;
}
```

suchipi commented on src/spec-jest/index.spec.js:
```diff
@@ -0,0 +1,25 @@
+import chai, { expect } from "chai";
```
This file is redundant, because it does the same thing as `examples/4_jestMode/spec.js`. Could you remove it?

In fact you can remove the whole spec-jest directory

suchipi commented on package.json:
```diff
@@ -6,12 +6,13 @@
   "scripts": {
     "build": "rm -rf dist/* && babel src --out-dir dist",
-    "mocha": "mocha -r 'dist/spec/index' 'dist/**/*.spec.js'",
+    "jest": "jest --roots dist/spec-jest",
```
We don't need this `jest` script here because it's handled by the example, so let's remove it.

suchipi commented on package.json:
```diff
@@ -6,12 +6,13 @@
     "example_2:watch": "npm run build && cd examples && npm run example_2:watch",
     "example_3:watch": "npm run build && cd examples && npm run example_3:watch",
-    "test": "npm run build && npm run mocha && npm run examples"
+    "test": "npm run build && npm run mocha && npm run jest && npm run examples"
```
We don't need to run jest here at the top level because it gets run by the examples, so let's remove it from the test script.

suchipi commented on package.json:
```diff
@@ -6,12 +6,13 @@
     "examples": "cd examples && npm test",
     "example_1:watch": "npm run build && cd examples && npm run example_1:watch",
     "example_2:watch": "npm run build && cd examples && npm run example_2:watch",
     "example_3:watch": "npm run build && cd examples && npm run example_3:watch",
```
Could you add a script for example_4:watch here like the other examples have?
```js
"example_4:watch": "npm run build && cd examples && npm run example_4:watch",
```

suchipi commented on package.json:
```diff
@@ -34,6 +35,7 @@
     "babel-preset-es2015": "6.9.0",
     "babel-preset-react": "6.11.1",
     "chai": "3.5.0",
+    "jest": "20.0.4",
```
Since jest isn't used by the project's tests, just the examples, could you move this devDependency to examples/package.json?

suchipi commented on README.md:
````diff
@@ -78,6 +78,17 @@ $ env CI=true npm test
+chai.use(chaiJestSnapshot);
+```
+
 ### Framework-agnostic Configuration Mode (Recommended for Non-Mocha Users)
````
Let's update this to say "Recommended for Non-Mocha/Jest Users"

suchipi commented on README.md:
````diff
@@ -78,6 +78,17 @@ $ env CI=true npm test
+```
+
 ### Framework-agnostic Configuration Mode (Recommended for Non-Mocha Users)
 If you are not using mocha as your test runner, it is recommended to use chai-jest-snapshot in "framework-agnostic configuration mode".
````
Let's update this to say "If you are not using mocha or jest as your test runner"

suchipi commented on src/buildMatchSnapshot.js:
```diff
@@ -33,4 +42,19 @@ const buildMatchSnapshot = (utils, parseArgs) => {
+
+const thisRunsInJest = () => (
+  typeof jest === "object" &&
+  JEST_MARKERS.every((marker) => typeof jest[marker] === "function")
```
@cpojer is there a more canonical way to detect that we're running in jest? Or is this fine?

suchipi replied:
(JEST_MARKERS is:
```
const JEST_MARKERS = ["enableAutomock", "genMockFromModule", "clearAllMocks", "runAllTicks"];
```
)

tdd commented:
Hey @suchipi I just pushed a commit addressing all your concerns.

As for canonical Jest detection, I couldn't find anything on Google or Jest's site to begin with, hence this home-baked heuristic. I intentionally focused on methods not likely to be found in other test runners (plus, the object is "jest", so…).

suchipi commented:
Looks great, thank you!

suchipi commented:
I will pull this and confirm it works on my machine, then merge and cut a version.

suchipi commented:
This has been published in v1.3.0. Thank you for the contribution! 🎉

tdd commented:
'most welcome!


On Fri, Jul 14, 2017, at 09:18, Stephen Scott wrote:
> This has been published in v1.3.0. Thank you for the contribution! 🎉> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub[1], or mute the thread[2]. > 

--
Christophe Porteneuve
tdd@tddsworld.com

Links:

  1. https://github.com/suchipi/chai-jest-snapshot/pull/23#issuecomment-315289789
  2. https://github.com/notifications/unsubscribe-auth/AAAOqLbRTXNFm_z26krPi__z_hDl8iNUks5sNxZSgaJpZM4OSmpE
